cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs

```diff
@@ -0,0 +1,353 @@
+import qualified Data.Set as Set
+import           Data.Time.Clock (DiffTime)
+
+import           Control.Tracer (Tracer, traceWith)
+import qualified Ouroboros.Consensus.Mempool as Mempool
+import qualified Ouroboros.Network.Protocol.TxSubmission.Type as TxSubmit
+
+-- | Bulk submisson of transactions.
+--
+--   This is intended to be used as a seperate process that interfaces
+--   between some form of transaction generation mechanism
+--   (e.g. benchmarking) and the local Ouroboros protocol peer. It has
+--   configurable latency hiding (number of transactions to try to
+--   keep ahead of actual submission) and has tracing that both
+--   reports observables related to transaction submission and whether
+--   the tx generation is able to keep up the pace.
+--
+--   For the details of the protocol description see
+--   `TxSubmit.TxSubmission` and its associated instances
+--   (e.g. `Message`).
+bulkSubmission
+  :: forall m blk txid tx .
+     ( MonadIO m, MonadSTM m, MonadTime m, MonadTimer m
+     , Ord txid
+     , Mempool.ApplyTx blk
+     , tx ~ Mempool.GenTx blk, txid ~ Mempool.GenTxId blk)
+  => (ROEnv txid tx -> ROEnv txid tx)
+  -- changes to default settings
+  -> Tracer m (TraceBenchTxSubmit txid)
+  -> LazyTVar m Bool
+  -- Set to True to indicate subsystem should terminate
+  -> LazyTMVar m [tx]
+  -- non-empty list of transactions to be forwarded,
+  -- empty list indicates terminating
+  -> LazyTMVar m (RPCTxSubmission m txid tx)
+  -- the RPC variable shared with
+  -- `TxSubmit.TxSubmission` local peer
+  -> m ()
+bulkSubmission updEnv tr termVar txIn rpcIn = do
+--    liftIO $ putStrLn "bulkSubmission__0"
+  defaultRWEnv >>= evalStateT (go $ updEnv defaultROEnv)
+ where
+  go :: ROEnv txid tx -> StateT (RWEnv m txid tx) m ()
+  go env = do
+    -- lift . traceWith tr . TraceBenchTxSubDebug $ "go, env: " ++ show env
+    shutdown <- gets (\RWEnv{terminating, inFlight, notYetSent} ->
+                       terminating &&
+                       Seq.null inFlight &&
+                       Seq.null notYetSent)
+    if shutdown
+    then do -- terminating normally, inform local peer
+      -- check if remote peer is waiting on us
+      -- lift . traceWith tr . TraceBenchTxSubDebug $ "TERMINATING normally"
+      (opP, op) <- maybe (False, error "internal error")
+                         (\x -> (True,snd x)) <$> gets availableOp
+--        liftIO $ putStrLn $ "TERMINATING normally, opP: " ++ show opP
+      lift . atomically $
+        if opP
+        then putTMVar op Nothing
+        else do
+          reqTxIds <- takeTMVar rpcIn -- no other message should occur as nothing is in flight
+          case reqTxIds of
+            RPCRequestTxIds _ resp -> putTMVar resp Nothing
+            _ -> return ()
+    else do -- process next interaction
+      -- lift . traceWith tr . TraceBenchTxSubDebug $ "go, process next interaction"
+      go1 env
+
+  go1 :: ROEnv txid tx -> StateT (RWEnv m txid tx) m ()
+  go1 env = do
```

Comment:
This function feels like it could greatly benefit from being split into parts, at least as a pipeline.

Reply:
This code was written by Neil. To be honest, I don't want to touch it right now (actually, only Neil really knows how this code works).